### PR TITLE
[FLINK-4505][Cluster Management] Implement TaskManagerFactory to brin…

### PR DESCRIPTION
**Pull request by wangzhijiang999**
Implement `TaskManagerRunner` to provide `createAndStartComponents`  static method for constructing related components of `TaskManager` and starting them in yarn or standalone mode.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/YarnTaskExecutorFactory.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	 * and select the one where the connection attempt is successful.
+	 * <p/>
+	 * After selecting the network interface, this method brings up an actor system
+	 * for the TaskManager and its actors, starts the TaskManager's services
+	 * (library cache, shuffle network stack, ...), and starts the TaskManager itself.
+	 *
+	 * @param configuration    The configuration for the TaskManager.
+	 * @param resourceID       The id of the resource which the task manager will run on.
+	 */
+	private TaskExecutor selectNetworkInterfaceAndRunTaskManager(
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/YarnTaskExecutorFactory.java]: Why does only the Yarn factory selects the network interface and not the standalone implementation?

**Comment by wangzhijiang999** [comment]
@mxm ,  The {{StandaloneTaskExecutorFactory}} can be used for mini cluster or testing mode I think, and the {{YarnTaskExecutorFactory}} used for yarn mode. After you confirm the implementation, I can add some testings for the factory if needed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/YarnTaskExecutorFactory.java (hunk 1 above)
Comment (wangzhijiang999) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/YarnTaskExecutorFactory.java]: The parameters provided from different modes are different as I referred to previous implementation for **TaskManager**. For example, **YarnTaskManagerRunner** invokes the method "selectNetworkInterfaceAndRunTaskManager", and **ForkableFlinkMiniCluster** invokes the method "startTaskManagerComponentsAndActor" to start **TaskManager** before. So I retained the previous ways for different modes.

**Comment by tillrohrmann** [comment]
Thanks for the contribution @wangzhijiang999.

I think the abstraction is not right. The `StandaloneFactory` and the `YarnFactory` should actually only differ in the class of the `TaskExecutor` they start. Furthermore, I think that the network interface selection should not be part of the factory. Instead I would pass the hostname and port to the `createTaskManager(hostname, port)` method which constructs the `TaskManager`. Maybe it makes even sense to pass the `RpcService` via the `createTaskManager(RpcService)` and defer the RpcService creation to the outside method running the select interface method. 

The difference between the Standalone/Yarn and Testing factory is that the first two create the TM components whereas the `TestingFactory` is initialized with the `TaskManager` components.

**Comment by wangzhijiang999** [comment]
@tillrohrmann , the current implementation is just follow the previous process for yarn and standalone modes. And I agree your opinion and actually it is not very clear for current ways. I think first we should re-define and confirm the parameters passed to the "createTaskManager" method for different factories, then I can fix the inner processes based on the input parameters. Would you suggest the specific parameters passed for different factories?  Thank you!

**Comment by tillrohrmann** [comment]
I think we don't need the distinction between standalone and yarn in the TaskManager startup. They should actually be the same. The difference will be the `StandaloneRunner`, `YarnRunner` and `MesosRunner`.

I think the factory should be responsible for creating the `TaskManager` components. I would pull out the `RpcService` creation and, thus, also the network interface selection, because this is code which can be re-used across different components (TM and JM, for example).

**Comment by wangzhijiang999** [comment]
@tillrohrmann , I tried to understand your idea as follows:
1. Provide specific TaskExecutorFactory class instead of abstract factory for both standalone/yarn mode.
2. Network selection and RPC service creation methods should be pulled out from factory, maybe remove to some utility class?(for JM reuse)
3. The constructor of TaskExecutorFactory supports many different parameters: such as 
TaskExecutorFactory(Configuration, ResourceID)
TaskExecutorFactory(RpcService, HighAvailabilityServices)
TaskExecutorFactory(hostname, port)
TaskExecutorFactory(Configuration, ResourceID, RpcService, HighAvailabilityServices,hostname, port), 
The above three constructors for partial parameters can be transformed into the fourth final constructor , all the missing parameters can be generated by internal default value.
4. The TaskExecutorFactory supports the method "createTaskManger" to bring up TaskManger for outside  world, and this method will construct the related components(TaskManagerConfiguration, NetworkManager, IOManager, MemoryManager)
5. For testing mode, construct the TestingTaskExecutorFactory to pass all the components explicitly, including (ResourceID,MemoryManager,IOManager,NetworkEnvironment,numberOfSlots, RpcService,HighAvailabilityServices), the TaskManagerConfiguration should be passed from outside or generate implicitly?
6. In addition, the localTaskManagerCommunication parameter is needed before to decide whether to create NettyConfig for standalone or yarn mode. Now I will remove this parameter to create ~~NettyConfig~~ always.

**Comment by tillrohrmann** [comment]
Hi @wangzhijiang999, maybe we should start simple without introducing a factory method, because there might actually be not many cases to distinguish. Maybe we could rename the `TaskManagerFactory` into `TaskManagerRunner` which has static methods to create the `TaskManagers` components and does the network selection. That way we keep the initialization and the actual `TaskManager` logic separated.

For testing purposes I guess we don't need to setup any components because they are usually mocked or one is using testing components. Passing these components to the constructor of the `TaskManager` should not be a big deal.

Does this make sense?

**Comment by wangzhijiang999** [comment]
Yes, that also make sense. For testing purpose it is very clear, so we do not need do anything currently, all the components in `TaskManager` constructor can be mocked implicitly.

For `TaskManagerRunner`, the purpose is to pull out the initialization of related components from `TaskManager` to make it logic clear. Just one issue to be confirmed, we should provide more static methods of different parameter units for outside world or just one static method such as 'selectNetworkInterfaceAndRunTaskManager(`Configuration` configuration,`ResourceID` resourceID) '? 
I think providing more methods with different parameters may be reasonable, because some parameters such as 'hostname','port', `RpcService`, `HighAvailabilityServices` may want to be passed from outside.

**Comment by tillrohrmann** [comment]
I think you should have at least another method `startComponents` which starts the different components. Everything else can be added later when we see that it would be helpful.

**Comment by wangzhijiang999** [comment]
@tillrohrmann , I created the `TaskExecutorRunner` for constructing the related components for starting `TaskExecutor`, and removed the factory class. Currently only the `ResourceID` and 'Configuration` parameters must be passed to the method `startComponents`, other parameters are optional. After you confirm the way , I will write some testings based on it. 
The `MemoryLogger` is removed from `TaskExecutorRunner` temporarily. The `ActorGateway` should not be passed to the constructor of `MemoryLogger` directly I think. If you agree, I will re-structure the `MemoryLogger` in another jira.

**Comment by tillrohrmann** [comment]
Thanks for your contribution @wangzhijiang999. I'll merge it.

**Comment by tillrohrmann** [comment]
I've merged your PR @wangzhijiang999. Thanks for your work :-) You can close this PR now.

**Comment by wangzhijiang999** [comment]
@tillrohrmann , thank you for merging and help. If there are any following works to do related with TaskManager, you can assign to me and I am willing to do.

**Pull request closed without merging**